File: C++/Tests/Question_1_1.cpp
#include <string>
#include <gtest/gtest.h>

/**
 * Question 1.1:
 * Implement an algorithm to determine if a string has all unique characters.
 * What if you cannot use additional data structures?
 */

/**
 * The solution will assume that the string will have only ASCII characters (256 valid characters)
 * @param  text The string text that will be checked.
 * @return      True if all the characters are unique in the string, false otherwise.
 */
bool HasUniqueChars(const std::string &text)
{
    // If a string of more than 256 characters is received on the input it means that at least
    // one character must be repeated.
    if (text.size() > 256)
        return false;

    // Considering that only ASCII characters are going to be received on the string
    // an array of 256 positions is enough to tell if there are any repeated chars
    // in the input string. An array of booleans will be used and it will be initialized
    // to false in all the positions.
    bool charArray[256] = {};

    // Take each letter of the input text and mark the position on the array for that
    // specific letter, if the position was already marked it means that a duplicated
    // character was found, access the chars by reference to avoid copying them.
    for (auto &c : text)
    {
        if (charArray[static_cast<int>(c)] == true)
            return false;
        charArray[static_cast<int>(c)] = true;
    }

    // If there were not repeated characters, return true to indicate that there are no duplicates.
    return true;
}

//! Test for non repeated characters.
TEST(Question_1_1, NonRepeatedChars)
{
    std::string text{ "abcdefg" };
    ASSERT_TRUE(HasUniqueChars(text));
}

//! Test for repeated characters.
TEST(Question_1_1, RepeatedChars)
{
    std::string text{ "abcdefgabcdefg" };
    ASSERT_FALSE(HasUniqueChars(text));
}

//! Test for empty string.
TEST(Question_1_1, EmptyInputString)
{
    std::string text{};
    ASSERT_TRUE(HasUniqueChars(text));
}

//! Test for a really long text
TEST(Question_1_1, LongInputString)
{
    std::string text{"abcdefghijklmnopqrstuvwxyz0123456789ß´ü+öä#,.-!§$%&()=?`*';:_{[]}~@€\
abcdefghijklmnopqrstuvwxyz0123456789ß´ü+öä#,.-!§$%&()=?`*';:_{[]}~@€\
abcdefghijklmnopqrstuvwxyz0123456789ß´ü+öä#,.-!§$%&()=?`*';:_{[]}~@€\
abcdefghijklmnopqrstuvwxyz0123456789ß´ü+öä#,.-!§$%&()=?`*';:_{[]}~@€"};
    ASSERT_FALSE(HasUniqueChars(text));
}

File: CSharp/Tests/Question_1_1.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests
{
    [TestClass]
    public class Question_1_1
    {
        // Question 1.1:
        // Implement an algorithm to determine if a string has all unique characters.
        // What if you cannot use additional data structures?

        // The solution will assume that only ASCII characters are going to be received.
        public bool HasUniqueChars(string text)
        {
            // If more than 256 characters in the string, then at least one is repeated.
            if (text.Length > 256)
                return false;

            // Array of booleans to check is a character is duplicated.
            var charArray = new bool[256];

            // Get each character from the string
            foreach (var c in text)
            {
                // If the character was already found, return false.
                if (charArray[c])
                    return false;
                // Mark the position on the array.
                charArray[c] = true;
            }

            // No repeated characters found, return true.
            return true;
        }

        // Test methods to ensure the solution was correct, will be using the unit testing
        // capabilities of the C# language.

        // No repeated characters in the text.
        [TestMethod]
        public void NonRepeatedChars()
        {
            const string text = "abcdefg";
            Assert.IsTrue(HasUniqueChars(text));
        }

        // Repeated characters in the text
        [TestMethod]
        public void RepeatedChars()
        {
            const string text = "abcdefgabcdefg";
            Assert.IsFalse(HasUniqueChars(text));
        }

        [TestMethod]
        public void EmptyInputString()
        {
            const string text = "";
            Assert.IsTrue(HasUniqueChars(text));
        }
    }
}
